README
==================
nautilus-pdf-tools
==================

A set of tools to work with PDF documents from Nautilus

For more information, please visit our website:

https://github.com/atareao/nautilus-pdf-tools
or
http://www.atareao.es

============
Requirements
============

Required dependencies:

python,
python-gi,
gir1.2-gtk-3.0,
gir1.2-gdkpixbuf-2.0,
gir1.2-nautilus-3.0,
python-nautilus,
pdf-tools

========
Download
========

Download the package:

$ git clone https://github.com/atareao/nautilus-pdf-tools.git


debian/changelog
nautilus-pdf-tools (1.2.9-0extras22.04.01) jammy; urgency=medium

  * Improvements to create nemo and caja

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 20 Aug 2022 20:00:46 +0200

nautilus-pdf-tools (1.2.9-0extras20.04.03) focal; urgency=medium

  * Change sh by plumbum

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 06 May 2020 20:00:46 +0200

nautilus-pdf-tools (1.2.9-0extras19.10.01) eoan; urgency=medium

  * Ported to Python3

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 22 Oct 2019 08:12:20 +0200

nautilus-pdf-tools (1.2.9-0extras19.06.7) disco; urgency=medium

  * Added dependencies for python3 used by Nemo
  * Fix minor issues
  * Change progress dialog

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 16 Oct 2019 06:05:42 +0200

nautilus-pdf-tools (1.2.9-0extras19.06.01) disco; urgency=medium

  * New vy to new version

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 14 Oct 2019 00:11:50 +0200

nautilus-pdf-tools (1.2.0-0extras19.04.1) disco; urgency=medium

  * add dependecy
  * fixed a bug with menu initialization GObject

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Fri, 27 Sep 2019 19:54:57 +0200

nautilus-pdf-tools (1.2.0-0extras16.04.0) xenial; urgency=medium

  * Sign
  * Dependencies
  * Encrypt and decrypt

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Fri, 30 Mar 2018 21:04:43 +0200

nautilus-pdf-tools (1.0.0-0extras16.04.1) xenial; urgency=medium

  * Progression
  * Added thumbnails for pdf and images
  * window a dependency
  * Updated spanish translations

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 21 Mar 2018 20:56:15 +0100

nautilus-pdf-tools (0.9.0-0extras16.04.7) xenial; urgency=medium

  * splitted

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 20 Mar 2018 20:32:58 +0100

nautilus-pdf-tools (0.8.0-0extras15.04.9) vivid; urgency=medium

  * Fixed minor bug

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 06 Feb 2016 11:19:21 +0100

nautilus-pdf-tools (0.8.0-0extras15.04.0) vivid; urgency=medium

  * Added drag&drop feature to join pdfs
  * Added drag&drop feature to convert images to pdfs
  * Supported a lot of image formats
  * Fixed flip bug
  * Rewrited the code to separate dialogs from main code
  * Watermark with a lot of image formats

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 17 Oct 2015 21:25:23 +0200

nautilus-pdf-tools (0.7.5-0extras15.04.0) vivid; urgency=medium

  * From a lot of image formats to PDF

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 17 Oct 2015 11:01:23 +0200

nautilus-pdf-tools (0.7.4-0extras15.04.0) vivid; urgency=medium

  * Added new option to convert pdf to png

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 04 Oct 2015 14:01:53 +0200

nautilus-pdf-tools (0.7.3-0extras15.04.0) vivid; urgency=medium

  * Added new option to create pdf file from images

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 04 Oct 2015 08:21:23 +0200

nautilus-pdf-tools (0.7.2-0extras15.04.0) vivid; urgency=medium

  * Fixed quote strings
  * Added new option to reduce pdfs

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 03 Oct 2015 20:14:57 +0200

nautilus-pdf-tools (0.7.1-0extras15.04.0) vivid; urgency=medium

  * Fixed extensions bug

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 27 Sep 2015 21:46:29 +0200

nautilus-pdf-tools (0.7.1-0extras14.04.1) trusty; urgency=low

  * Migrated to Trusty Tahr
  * Fixed a bug with convert extensions

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 09 Mar 2014 21:23:16 +0100

nautilus-pdf-tools (0.7.1-0extras13.10.3) saucy; urgency=low

  * Fixed a bug with textmark

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 30 Dec 2013 08:43:52 +0100

nautilus-pdf-tools (0.7.1-0extras13.10.2) saucy; urgency=low

  * Fixed a bug with dependencies

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 29 Dec 2013 13:03:58 +0100

nautilus-pdf-tools (0.7.1-0extras13.10.0) saucy; urgency=low

  * Remove PyPdf all dependencies. All manipulation is done with cairo.

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 29 Dec 2013 12:23:16 +0100

nautilus-pdf-tools (0.7.0-0extras13.10.0) saucy; urgency=low

  * Added new feature to flip pdfs
  * Added a dialog to show flip and rotate operations
  * Using cairo to rotate and flip pdfs
  * Added a new feature to watermark pdfs with png images

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 28 Dec 2013 20:22:44 +0100

nautilus-pdf-tools (0.6.4-0extras13.10.0) saucy; urgency=low

  * Ported to saucy

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 20 Nov 2013 12:30:23 +0100

nautilus-pdf-tools (0.6.3-0extras13.04.2) raring; urgency=low

  * Ported to raring

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 02 Jun 2013 23:06:43 +0200

nautilus-pdf-tools (0.0.6.2-1ubuntu1) precise; urgency=low

  * Added new paper sizes

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 14 Aug 2012 10:33:13 +0200

nautilus-pdf-tools (0.0.6.1-1ubuntu1) precise; urgency=low

  * Fixed a bug with extract text feature

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 31 Mar 2012 20:44:25 +0200

nautilus-pdf-tools (0.0.6.0-1ubuntu1) precise; urgency=low

  * Modified resize feature
  * Added extract text feature

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 31 Mar 2012 19:30:59 +0200

nautilus-pdf-tools (0.0.5.0-1ubuntu1) precise; urgency=low

  * Added resize feature
  * Added spanish translation

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 25 Mar 2012 19:10:16 +0200

nautilus-pdf-tools (0.0.4.0-1ubuntu1) precise; urgency=low

  * Added combination feature

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 25 Mar 2012 12:24:59 +0200


src/pdf-tools.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright (c) 2012 Lorenzo Carbonell Cerezo <a.k.a. atareao>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import gi
try:
    gi.require_version('Nautilus', '3.0')
    gi.require_version('GObject', '2.0')
    gi.require_version('Gtk', '3.0')
    gi.require_version('GdkPixbuf', '2.0')
except ValueError as e:
    print(e)
    exit(1)
from gi.repository import GdkPixbuf, GObject, Gtk
from gi.repository import Nautilus as FileManager
import sys

try:
    sys.path.insert(0, '/usr/share/nautilus-python/extensions/pdf-tools/')
    import pdfmanager
    from comun import APPNAME, ICON, VERSION
    from comun import _
    import tools
except Exception as e:
    print(e)
    exit(-1)

SEPARATOR = u'\u2015' * 10

"""
Tools to manipulate pdf
"""


class PdfToolsMenuProvider(GObject.GObject, FileManager.MenuProvider):
    """Implements the 'Replace in Filenames' extension to the File Manager
    right-click menu"""

    def __init__(self):
        """File Manager crashes if a plugin doesn't implement the __init__
        method"""
        GObject.Object.__init__(self)

    def doit(self, menu, option, selected, window):
        if option == 'rotate' or option == 'watermark' or \
                option == 'textmark' or option == 'paginate' or \
                option == 'sign':
            pdfmanager.operate(option, selected, window)
        elif option == 'remove pages':
            pdfmanager.remove_some_pages(selected, window)
        elif option == 'extract pages':
            pdfmanager.extract_some_pages(selected, window)
        elif option == 'join':
            pdfmanager.join_pdf_files(selected, window)
        elif option == 'split':
            pdfmanager.split_pdf_files(selected, window)
        elif option == 'combine':
            pdfmanager.combine_pdf_pages(selected, window)
        elif option == 'reduce':
            pdfmanager.reduce(selected, window)
        elif option == 'resize':
            pdfmanager.resize_pdf_pages(selected, window)
        elif option == 'convert2png':
            pdfmanager.convert_pdf_file_to_png(selected, window)
        elif option == 'convert2pdf':
            pdfmanager.create_pdf_from_images(selected, window)
        elif option == 'encrypt':
            pdfmanager.encrypt_files(selected, window)
        elif option == 'decrypt':
            pdfmanager.decrypt_files(selected, window)

    def get_file_items(self, window, sel_items):
        """Adds the 'Replace in Filenames' menu item to the File Manager
        right-click menu, connects its 'activate' signal to the 'run' method
        passing the selected Directory/File"""
        if tools.all_files_are_pdf(sel_items):
            top_menuitem = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-pdf-tools',
                label=_('Pdf Tools'),
                tip=_('Tools to manipulate pdf files'),
                icon='Gtk-find-and-replace')
            submenu = FileManager.Menu()
            top_menuitem.set_submenu(submenu)
            if len(sel_items) == 1:
                items = [
                    ('01', _('Rotate and flip'),
                     _('rotate_and_flip pdf files'),
                     'rotate'),
                    ('02', _('Watermark'), _('Watermark pdf files'),
                     'watermark'),
                    ('03', _('Textmark'), _('Textmark pdf files'),
                     'textmark'),
                    ('04', _('Paginate'), _('Paginate pdf files'),
                     'paginate'),
                    ('05', _('Sign'), _('Sign pdf files'),
                     'sign'),
                ]
            else:
                items = []
            moreitems = [
                ('06', _('Remove pages'),
                 _('Remove pages of the document files'),
                 'remove pages'),
                ('07', _('Extract pages'),
                 _('Extract pages of the document files'),
                 'extract pages'),
                ('08', _('Join pdf files'),
                 _('Join pdf files in one document'),
                 'join'),
                ('09', _('Split pdf files'),
                 _('Split a pdf in several documents'),
                 'split'),
                ('10', _('Combine pdf pages'),
                 _('Combine pdf pages in one page'),
                 'combine'),
                ('11', _('Reduce pdf size'), _('Reduce pdf size'),
                 'reduce'),
                ('12', _('Resize pdf pages'), _('Resize pdf pages'),
                 'resize'),
                ('13', _('Convert pdf to png'),
                 _('Convert pdf file to png image'),
                 'convert2png'),
                ('14', _('Encrypt'),
                 _('Encrypt pdf files'),
                 'encrypt'),
                ('15', _('Decrypt'),
                 _('Decrypt pdf files'),
                 'decrypt'),
            ]
            items = items + moreitems
            for item in items:
                sub_menuitem = FileManager.MenuItem(
                    name='PdfToolsMenuProvider::Gtk-pdf-tools-' + item[0],
                    label=item[1],
                    tip=item[2])
                sub_menuitem.connect('activate',
                                     self.doit,
                                     item[3],
                                     sel_items,
                                     window)
                submenu.append_item(sub_menuitem)

            sub_menuitem_98 = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-None',
                label=SEPARATOR)
            submenu.append_item(sub_menuitem_98)

            sub_menuitem_99 = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-pdf-tools-99',
                label=_('About'),
                tip=_('About'),
                icon='Gtk-find-and-replace')
            sub_menuitem_99.connect('activate', about)
            submenu.append_item(sub_menuitem_99)

            return top_menuitem,
        elif tools.all_files_are_images(sel_items):
            top_menuitem = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-pdf-tools',
                label=_('Pdf Tools'),
                tip=_('Tools to manipulate pdf files'),
                icon='Gtk-find-and-replace')
            submenu = FileManager.Menu()
            top_menuitem.set_submenu(submenu)
            items = [
                ('51', _('Convert to pdf'),
                 _('Convert images to pdf'),
                 'convert2pdf')
            ]
            for item in items:
                sub_menuitem = FileManager.MenuItem(
                    name='PdfToolsMenuProvider::Gtk-pdf-tools-' + item[0],
                    label=item[1],
                    tip=item[2])
                sub_menuitem.connect('activate',
                                     self.doit, item[3], sel_items, window)
                submenu.append_item(sub_menuitem)

            sub_menuitem_98 = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-None',
                label=SEPARATOR)
            submenu.append_item(sub_menuitem_98)

            sub_menuitem_99 = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-pdf-tools-99',
                label=_('About'),
                tip=_('About'),
                icon='Gtk-find-and-replace')
            sub_menuitem_99.connect('activate', about)
            submenu.append_item(sub_menuitem_99)

            return top_menuitem,
        return


def about(widget):
    ad = Gtk.AboutDialog()
    ad.set_name(APPNAME)
    ad.set_version(VERSION)
    ad.set_copyright('Copyrignt (c) 2012\nLorenzo Carbonell')
    ad.set_comments(_('Pdf Tools'))
    ad.set_license('''
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
''')
    ad.set_website('https://www.atareao.es')
    ad.set_website_label('atareao.es')
    ad.set_authors([
        'Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>'])
    ad.set_documenters([
        'Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>'])
    ad.set_logo(GdkPixbuf.Pixbuf.new_from_file(ICON))
    ad.set_icon(GdkPixbuf.Pixbuf.new_from_file(ICON))
    ad.set_program_name(APPNAME)
    ad.run()
    ad.destroy()
